client/tso_service_discovery.go

```diff
@@ -251,69 +426,129 @@ func (c *tsoServiceDiscovery) SetTSOGlobalServAddrUpdatedCallback(callback tsoGl
+		if err := c.globalAllocPrimariesUpdatedCb(newPrimary); err != nil {
 			return err
 		}
 	}
-	log.Info("[tso] switch primary", zap.String("new-primary", addr), zap.String("old-primary", oldPrimary))
+	log.Info("[tso] switch primary",
+		zap.String("new-primary", newPrimary),
+		zap.String("old-primary", oldPrimary))
 	return nil
 }
 
-func (c *tsoServiceDiscovery) updateMember() error {
-	resp, err := c.metacli.Get(c.ctx, []byte(c.primaryKey))
+func (c *tsoServiceDiscovery) updateMember() (err error) {
+	// The keyspace membership or the primary serving address of the keyspace group, to which this
+	// keyspace belongs, might have been changed. We need to query tso servers to get the latest info.
+	// If there are known healthy tso servers serving this keyspace, we start with them; otherwise,
+	// we start with any healthy tso server.
+	var keyspaceGroup *tsopb.KeyspaceGroup
+	tsoServerAddr := c.keyspaceGroupSD.getServer()
+	if len(tsoServerAddr) > 0 {
+		keyspaceGroup, err = c.findGroupByKeyspaceID(c.keyspaceID, tsoServerAddr, updateMemberTimeout)
+		c.keyspaceGroupSD.countFailure(err)
+	} else {
+		if tsoServerAddr, err = c.getTSOServer(c.apiSvcDiscovery); err != nil {
+			return err
+		}
+		keyspaceGroup, err = c.findGroupByKeyspaceID(c.keyspaceID, tsoServerAddr, updateMemberTimeout)
+		c.tsoServerDiscovery.countFailure(err)
+	}
 	if err != nil {
-		log.Error("[tso] failed to get the keyspace serving endpoint", zap.String("primary-key", c.primaryKey), errs.ZapError(err))
 		return err
 	}
 
-	if resp == nil || len(resp.Kvs) == 0 {
-		log.Error("[tso] didn't find the keyspace serving endpoint", zap.String("primary-key", c.primaryKey))
-		return errs.ErrClientGetServingEndpoint
-	} else if resp.Count > 1 {
-		return errs.ErrClientGetMultiResponse.FastGenByArgs(resp.Kvs)
+	// Initialize the all types of serving addresses from the returned keyspace group info.
+	primaryAddr := ""
+	secondaryAddrs := make([]string, 0)
+	addrs := make([]string, 0, len(keyspaceGroup.Members))
+	for _, m := range keyspaceGroup.Members {
+		addrs = append(addrs, m.Address)
+		if m.IsPrimary {
+			primaryAddr = m.Address
+		} else {
+			secondaryAddrs = append(secondaryAddrs, m.Address)
+		}
+	}
+
+	// If the primary address is not empty, we need to create a grpc connection to it, and do it
+	// out of the critical section of the keyspace group service discovery.
+	if len(primaryAddr) > 0 {
+		if primarySwitched := !strings.EqualFold(primaryAddr, c.getPrimaryAddr()); primarySwitched {
+			if _, err := c.GetOrCreateGRPCConn(primaryAddr); err != nil {
+				log.Warn("[tso] failed to connect the next primary",
+					zap.String("next-primary", primaryAddr), errs.ZapError(err))
+				return err
+			}
+		}
+	}
+
+	oldPrimary, primarySwitched := c.keyspaceGroupSD.update(keyspaceGroup, primaryAddr, secondaryAddrs, addrs)
+	if primarySwitched {
+		if err := c.afterPrimarySwitched(oldPrimary, primaryAddr); err != nil {
+			return err
+		}
+	}
+
+	if len(primaryAddr) == 0 {
```

Comment:
Shouldn't, because even there is no primary, which could happen, it still needs to update other info returned, including keyspace group info and secondaries info.

Reply:
Added comment
// Even if the primary address is empty, we still updated other returned info above, including the keyspace
// group info and the secondary addresses.